File: Trees/Python/Construct Binary Tree from Parent Array.py
# Problem link: https://www.geeksforgeeks.org/problems/construct-binary-tree-from-parent-array/1
# Time: O(n)
# Space: O(n) 

class Solution:
    def createTree(self,parent):
        store = {}
        root = -1
        
        for i in range(len(parent)):
            if parent[i] not in store:
                store[parent[i]] = Node(parent[i])
            
            if i not in store:
                store[i] = Node(i)
            
            if parent[i] == -1:
                root = i
            else:
                if not store[parent[i]].left:
                    store[parent[i]].left = store[i]
                else:
                    store[parent[i]].right = store[i]
                    
        return store[root]

File: Trees/Java/Construct Binary Tree from Parent Array.java
// Problem link: https://www.geeksforgeeks.org/problems/construct-binary-tree-from-parent-array/1
// Time: O(n)
// Space: O(n)
class Solution {
    public Node createTree(int parent[]) {
        Map<Integer, Node> store = new HashMap<>();
        int root = -1;
        
        for(int idx = 0; idx < parent.length; idx++) {
            store.putIfAbsent(parent[idx], new Node(parent[idx]));
            store.putIfAbsent(idx, new Node(idx));
            
            if(parent[idx] == -1) {
                root = idx;
            }
            else {
                if(store.get(parent[idx]).left == null) {
                    store.get(parent[idx]).left = store.get(idx);
                }
                else {
                    store.get(parent[idx]).right = store.get(idx);
                }
            }
        }
        
        return store.get(root);
    }
}
